### ISSUE TITLE
[Roadmap] PyTorch SparseTensor Support 🚀

### ISSUE THREAD

**Issue by EdisonLeeeee** [pyg-team/pytorch_geometric#5867]
### 🚀 The feature, motivation and pitch

# 🚀 The feature, motivation and pitch
PyG currently accepts `torch.LongTensor: edge_index` and `torch_sparse.SparseTensor: adj` in `MessagePassing`, which limits its flexibility for users who use native PyTorch SparseTensor. As such, an additional step is required for them to convert one to another.

The goal of this roadmap is to track the integration of native PyTorch SparseTensor support into PyG. After integration, the `edge_index` in `MessagePassing` can also accept PyTorch SparseTensor while allowing backpropagation like `torch_sparse.SparseTensor`.

## General Roadmap

- [ ] Add PyTorch SparseTensor support for `torch_geometric.nn.MessagePassing`
- [ ] Add PyTorch SparseTensor support for `torch_geometric.transforms`
- [ ] Add PyTorch SparseTensor support for `torch_geometric.loader`
- [ ] Replace `torch_sparse.SparseTensor` based implementations with PyTorch functionality
- [ ] ...

**Comment by EdisonLeeeee** [comment]
So far I can think of a way to support torch SparseTensor with the least effort in `MessagePassing` is:
```python
def hook(model, input):
    edge_index, size, kwargs = input
    if is_torch_coo_tensor(adj):
        edge_index = adj._indices()
        kwargs['edge_weight'] = adj._values()
    return edge_index, size, kwargs

layer.register_propagate_forward_pre_hook(hook)
``` 
we can register such a hook in `__init__` of `MessagePassing`, WDYT? @rusty1s

**Comment by rusty1s** [comment]
I think this would only be a temporary solution. Ultimately, we want direct support for it similar to `SparseTensor`. Wondering what are the main challenges to support this?

**Comment by EdisonLeeeee** [comment]
IMO, there are some challenges to support this:

1. It only supports `sum` aggregation when using PyTorch SparseTensor. For other cases including advanced aggregations, we still need to convert it to the form of `(edge_index, edge_weight)`.
2. PyTorch SparseTensor does not support `spspmm` or `mul` with broadcasting, which makes it challenging to implement some operations such as `gcn_norm` (e.g., $A\cdot D^{-\frac{1}{2}}$ ) efficiently.
3. As mentioned above, since PyTorch SparseTensor is also an instance of `torch.Tensor`, we should change the condition `isinstance(edge_index, Tensor)` if we want to support it. There should be some auxiliary functions like `is_edge_index`, `is_torch_coo_tensor`, or something else.

Please correct me if something is missing.

**Comment by rusty1s** [comment]
These are good thoughts. Appreciate it. Let me think.

1. The only way I can think of supporting this is to provide a `spmm` function in PyG that supports both `torch_sparse` and PyTorch. For PyTorch, we can error out in case `aggr != "sum"` or `aggr != "mean"`.
2. Yes, this is a real problem. We would need to provide our own implementation for this by working on the indices and values directly. Happy to move this to follow-up PRs and work on general `MessagePassing` integration first
3. Yes, this is expected, but IMO not a blocker, right?

**Comment by EdisonLeeeee** [comment]
You are right. Thanks for making it clear. Will update the roadmap correspondingly and make some PRs soon :)

**Comment by rusty1s** [comment]
Nice to have you back :)

**Comment by sandeep-189** [comment]
Hi, I was trying to use pytorch sparse for a RGCN like model that I was working on. I noticed that the mean aggregation for the pytorch sparse tensor implementation in the spmm function is raising a NotImplementedError. Should the implementation for this be torch.sparse.mm(src, other)/src.shape[0] ? Am I misunderstanding mean aggregation?

**Comment by rusty1s** [comment]
Mean aggregation would refer to a row-wise mean that only normalized across non-zero values. I don't think this is super hard to integrate though @EdisonLeeeee.

**Comment by EdisonLeeeee** [comment]
Yeah. Will take a look and make a PR for it.

**Comment by sandeep-189** [comment]
Thank you very much for the clarification. I think with that PR, RGCNConv and FastRGCNConv would also be working.

**Comment by EdisonLeeeee** [comment]
I've made a PR for it: #6868

**Comment by romankouz** [comment]
![image](https://github.com/pyg-team/pytorch_geometric/assets/58533981/0d102cb2-19b2-4f6a-9700-3d62b288db2b)

![image](https://github.com/pyg-team/pytorch_geometric/assets/58533981/a002c5ca-cd04-4126-9d16-9e42d9641458)

Apologies if I should have opened a new issue for this, but I opened #6889 right around the time the influx of `SparseTensor` support started and have some follow up questions.

Is the support for `SparseTensor` simply so that these method can accept a `SparseTensor` , or so that they use something like `spmm` for matrix multiplication? As far as I can see `GMMConv` still does not support sparse multiplication even if you pass in a sparse tensor. The commit you have for `GMMConv` `SparseTensor` support adds tests that allows you to pass `SparseTensor` but I don't think it does anything with the sparsity of the matrices. Is this a correct assessment? 

I ask not because I'm judging, but because I don't know if there is something I'm missing and just wanted to check :).

**Comment by rusty1s** [comment]
Hi @romankouz,

Note that this issue is more about bringing support of `torch.sparse.tensor` to PyG, not the `SparseTensor` class introduced by `torch-sparse` a while ago.

Overall, we aim to support as many layers as possible to accept sparse tensors as input, while only a fraction of them can actually make use of the `spmm` code path. As a general rule of thumb, we can only leverage `spmm` if messages are not dependent on destination node features and no edge features are involved (which is the case for `GMMConv`). The other main benefit you get from using sparse tensors is deterministic behavior since the aggregation kernels no longer rely on atomic operations.

**Comment by romankouz** [comment]
Hi @rusty1s ,

Thank you for the follow up! I'm not entirely sure how aggregation kernels rely on atomic operations for dense matrices but not for sparse tensors. However, I do realize my misunderstanding regarding SparseTensor integration. Thank you for the response!

**Comment by rusty1s** [comment]
The main difference between the two representations is that for `SparseTensor` it is guaranteed that neighborhoods are grouped contiguous in memory, which lets us utilize `segment_reduce` rather than `scatter_reduce` from `torch-sparse`. As such, this lets us leverage an alternative aggregation in which we no longer need to rely on atomic operations.
